netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```diff
@@ -586,12 +606,50 @@ private Http2Stream requireHttp2Stream(int streamId) {
     return stream;
   }
 
+  /**
+   * This is called exactly once.
+   */
+  void startDeadlineWatchdog(Runnable onTimeout, Deadline deadline) {
+    Preconditions.checkArgument(deadline != null, "deadline");
+    synchronized (lock) {
+      if (watchdogCanceled != null) {
+        // watchdog canceled already before it even started
+        Preconditions.checkState(watchdogCanceled, "watchdogCanceled");
+        return;
+      }
+      watchdogCanceled = Boolean.FALSE;
+      ScheduledExecutorService scheduler = SharedResourceHolder.get(timerService);
+      deadlineWatchdog = scheduler.schedule(
+          onTimeout, deadline.timeRemaining(TimeUnit.MILLISECONDS), TimeUnit.MILLISECONDS);
+      SharedResourceHolder.release(timerService, scheduler);
```

Comment:
FYI: this may not yield good results, since release is intended to be used after the resource is no longer being used. Here, the scheduler is still being used, we just don't reference it any more. This could cause extra threads (although wouldn't because ChannelImpl holds on to a reference) as the scheduler would be shut down, but still have things queued.